### PR TITLE
Use `HF_HUB_OFFLINE` + fix has_file in offline mode

### LINKED ISSUE #30345: Problem with pretrained Transformers in Offline mode
### System Info

For me, the usage of pretrained transformers in Offline mode is critical. Originally, the Module was OK when the needed files are available in "huggingface_hub" cache folder. 

But now it is not working anymore,, whatever the status of my cached folder is, the library performs at least one "HEAD" request which I don't want.

Is there a way to address this situation? And make the program conform to Online/Offline mode described in huggingface_hub documentation?

Regards, Marian

### Who can help?

_No response_

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

set HF_HUB_OFFLINE=1

model_checkpoint = "Davlan/bert-base-multilingual-cased-ner-hrl"
var tokenizer = AutoTokenizer.from_pretrained(model_checkpoint)

"""
this will cause:
  File "python-3.11.9-embed-amd64\Lib\site-packages\transformers\models\auto\tokenization_auto.py", line 804, in from_pretrained
    tokenizer_config = get_tokenizer_config(pretrained_model_name_or_path, **kwargs)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "python-3.11.9-embed-amd64\Lib\site-packages\transformers\models\auto\tokenization_auto.py", line 637, in get_tokenizer_config
    resolved_config_file = cached_file(
                           ^^^^^^^^^^^^
  File "python-3.11.9-embed-amd64\Lib\site-packages\transformers\utils\hub.py", line 398, in cached_file
    resolved_file = hf_hub_download(
                    ^^^^^^^^^^^^^^^^
  File "python-3.11.9-embed-amd64\Lib\site-packages\huggingface_hub\utils\_validators.py", line 119, in _inner_fn
    return fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^
  File "python-3.11.9-embed-amd64\Lib\site-packages\huggingface_hub\file_download.py", line 1261, in hf_hub_download
    metadata = get_hf_file_metadata(
               ^^^^^^^^^^^^^^^^^^^^^
  File "python-3.11.9-embed-amd64\Lib\site-packages\huggingface_hub\utils\_validators.py", line 119, in _inner_fn
    return fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^
  File "python-3.11.9-embed-amd64\Lib\site-packages\huggingface_hub\file_download.py", line 1674, in get_hf_file_metadata
    r = _request_wrapper(
        ^^^^^^^^^^^^^^^^^
  File "python-3.11.9-embed-amd64\Lib\site-packages\huggingface_hub\file_download.py", line 369, in _request_wrapper
    response = _request_wrapper(
               ^^^^^^^^^^^^^^^^^
  File "python-3.11.9-embed-amd64\Lib\site-packages\huggingface_hub\file_download.py", line 392, in _request_wrapper
    response = get_session().request(method=method, url=url, **params)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "python-3.11.9-embed-amd64\Lib\site-packages\requests\sessions.py", line 589, in request
    resp = self.send(prep, **send_kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "python-3.11.9-embed-amd64\Lib\site-packages\requests\sessions.py", line 703, in send
    r = adapter.send(request, **kwargs)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "python-3.11.9-embed-amd64\Lib\site-packages\requests\adapters.py", line 513, in send
    raise ProxyError(e, request=request)
requests.exceptions.ProxyError: HTTPSConnectionPool(host='huggingface.co', port=443): Max retries exceeded with url: /Davlan/bert-base-multilingual-cased-ner-hrl/resolve/main/tokenizer_config.json (Caused by ProxyError('Cannot connect to proxy.', NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000002A44F5A1250>: Failed to establish a new connection: [WinError 10061] Es konnte keine Verbindung hergestellt werden, da der Zielcomputer die Verbindung verweigerte')))
"""

### Expected behavior

The expection should not be raised. According to huggingface_hub documentation, the HF_HUB_OFFLINE should disable communication. Especially, by fact that the model files are precached by me into a local folder.

### PR DESCRIPTION
**Pull request by Wauplin**
Should fix https://github.com/huggingface/transformers/issues/30345.
Related to https://github.com/huggingface/transformers/pull/31010 as well, even though it makes sense to merge both PRs IMO.

This PR:
- delegates offline mode to `huggingface_hub` with `HF_HUB_OFFLINE` env variable instead of `TRANSFORMERS_OFFLINE`. This change is still backward-compatible meaning that users that had already set `TRANSFORMERS_OFFLINE` won't notice a difference. The goal is to have a single environment variable to disable network in all HF libraries.
- updates the docs/tests to showcase `HF_HUB_OFFLINE` env variable
- make sure to respect offline mode in `has_file` => it was not the case before, causing https://github.com/huggingface/transformers/issues/30345 and https://github.com/huggingface/transformers/pull/31004
- add `local_files_only` and `cache_dir` to `has_file()` helper. At the moment, if connection/network is disabled, an error is raised. But if an authentication/authorization problem occurs, it returns False. IMO this is not consistent and we should aim at having the same response if server can't be requested. Instead of returning `False`, I suggest we default to checking the cache directory in case the information exists there.
- adds a test to check `has_file` in offline mode

Please let me know what you think.

(Note: some tests are failing but I prefer to get theoretical approval before moving forward on this).

### CODE REVIEW

**Review by amyeroberts — COMMENTED** [review]
Thanks for working on this! 

Overall I think it looks good - only thing is that I think we need a deprecation cycle for `TRANSFORMERS_OFFLINE`

File: tests/utils/test_hub_utils.py [hunk 1]
```diff
@@ -99,9 +101,20 @@ def test_non_existence_is_cached(self):
...
+    def test_has_file_in_cache(self):
+        with tempfile.TemporaryDirectory() as tmp_dir:
+            # Empty cache dir + offline mode => return False
+            assert not has_file(TINY_BERT_PT_ONLY, WEIGHTS_NAME, local_files_only=True, cache_dir=tmp_dir)
+
+            # Populate cache dir
+            hf_hub_download(TINY_BERT_PT_ONLY, WEIGHTS_NAME, cache_dir=tmp_dir)
+
+            # Cache dir + offline mode => return True
+            assert has_file(TINY_BERT_PT_ONLY, WEIGHTS_NAME, cache_dir=tmp_dir)
```
[the remark this answers is not in the document]
Comment (Wauplin) [reply tests/utils/test_hub_utils.py]: Yes! Fixed in https://github.com/huggingface/transformers/pull/31016/commits/eb5968931fe1b0fa09befb98dcb9fe474a8acc71.

**Comment by amyeroberts** [comment]
> Agree but if we want to deprecate it, I think we should do it in huggingface_hub directly.

OK - sounds good to me! 

> In any case I don't think we'll remove it before transformers v5, right?

I don't know, tbh, probably not before v5 as it's a common argument but if we think that's going to be too far away we could always move if forward

**Comment by marianpascalau** [comment]
> Should fix #30345. Related to #31010 as well, even though it makes sense to merge both PRs IMO.
> 
> This PR:
> 
> * delegates offline mode to `huggingface_hub` with `HF_HUB_OFFLINE` env variable instead of `TRANSFORMERS_OFFLINE`. This change is still backward-compatible meaning that users that had already set `TRANSFORMERS_OFFLINE` won't notice a difference. The goal is to have a single environment variable to disable network in all HF libraries.
> * updates the docs/tests to showcase `HF_HUB_OFFLINE` env variable
> * make sure to respect offline mode in `has_file` => it was not the case before, causing [Problem with pretrained Transformers in Offline mode  #30345](https://github.com/huggingface/transformers/issues/30345) and [Do not trigger autoconversion if local_files_only #31004](https://github.com/huggingface/transformers/pull/31004)
> * add `local_files_only` and `cache_dir` to `has_file()` helper. At the moment, if connection/network is disabled, an error is raised. But if an authentication/authorization problem occurs, it returns False. IMO this is not consistent and we should aim at having the same response if server can't be requested. Instead of returning `False`, I suggest we default to checking the cache directory in case the information exists there.
> * adds a test to check `has_file` in offline mode
> 
> Please let me know what you think.
> 
> (Note: some tests are failing but I prefer to get theoretical approval before moving forward on this).

I want to make a suggestion. In order to better follow the implementation in "huggingface_hub", instead of following the envVariable you better follow the boolean variable (with the same name) from withing huggingface_hub module:

```
from huggingface_hub.constants import HF_HUB_OFFLINE

if HF_HUB_OFFLINE:
    print("Now we are offline")

```
This will better follow the huggingface implementation when they decide to change something.

**Comment by Wauplin** [comment]
Hi @marianpascalau, thanks for the suggestion! Yes, that's exactly what I'm doing [here](https://github.com/huggingface/transformers/pull/31016/files#diff-82b93b530be62e40679876a764438660dedcd9cc9e33c2374ed21b14ebef5dbaR80) to delegate the env variable handling to `huggingface_hub`.

**Comment by ydshieh** [comment]
Hi @Wauplin Is this PR also trying to fix the currently failing tests seen in other PRs too?

> python3 -m pytest -v tests/utils/test_offline.py::OfflineTests::test_offline_mode

**Comment by Wauplin** [comment]
Hi @ydshieh I think https://github.com/huggingface/transformers/pull/31010 should fix it (but maybe this one as well). Both are kinda related

**Comment by ydshieh** [comment]
Thanks! I will check

**Review by LysandreJik — APPROVED** [review]
Thanks @Wauplin ! To ensure that the deprecation cycle is respected, could we keep a test with `TRANSFORMERS_OFFLINE` while ensuring that it throws a warning?

As @amyeroberts said, this was a very well advertised argument

**Comment by Wauplin** [comment]
Yes @LysandreJik @amyeroberts completly right. I reverted the tests to use `TRANSFORMERS_OFFLINE` (then no breaking change compared to `main`). And I added one to test that `HF_HUB_OFFLINE` is respected as well.

**Review by amyeroberts — APPROVED** [review]
Thanks for handling this!

**Comment by Wauplin** [comment]
@amyeroberts Tests are passing now. Feel free to do a last review or merge when it's ready :)

**Comment by 311-code** [comment]
I would also like to see added to hub.py's `def send_telemetry` something like:

`    def redact_sensitive_info(value): # Redact sensitive information such as paths or tokens if isinstance(value, str) and (os.path.exists(value) or 'token' in value.lower()): return "[REDACTED]" return value`

because how it is now seems like a bit of a security issue to me with the:

```
        args_as_dict = {k: v for k, v in args.__dict__.items() if not k.startswith("_") and v is not None}
        if "model_name_or_path" in args_as_dict:
            model_name = args_as_dict["model_name_or_path"]
```

and        ```
 if "dataset_name" in args_as_dict:
            data["dataset_name"] = args_as_dict["dataset_name"]
```
I noticed there were some bert models in 2022 trained on imdb and really hope this is not collecting training data on celebrity strings... It only seems to filter _ and None.

**Comment by Wauplin** [comment]
Hi @311-code, concerns are legit but could you open a new issue on Github for it? This PR is related to offline mode only (and just so you know, when offline mode is enabled, telemetry is automatically stopped)

**Comment by 311-code** [comment]
Yes i won't be in front of PC until late tonight but I will highlight things I noticed. I am not quite sure if the redact code will work or have side effects yet.

**Pull request merged**